src/org/rust/lang/icons/RustIconProvider.kt

```diff
@@ -1,17 +1,60 @@
 package org.rust.lang.icons
 
 import com.intellij.ide.IconProvider
+import com.intellij.openapi.util.Iconable
 import com.intellij.psi.PsiElement
-import org.rust.lang.core.psi.impl.RustFnItemImpl
-import org.rust.lang.core.psi.impl.RustImplMethodImpl
+import com.intellij.psi.util.PsiUtil
+import com.intellij.ui.RowIcon
+import com.intellij.util.PlatformIcons
+import com.intellij.util.VisibilityIcons
+import com.intellij.util.ui.EmptyIcon
+import org.rust.lang.core.psi.*
 import javax.swing.Icon
 
 class RustIconProvider: IconProvider() {
     override fun getIcon(element: PsiElement, flags: Int): Icon? {
         return when (element) {
-            is RustFnItemImpl -> RustIcons.FUNCTION
-            is RustImplMethodImpl -> RustIcons.METHOD
+            is RustEnumItem -> getEnumIcon(element, flags)
+            is RustEnumDef -> RustIcons.FIELD
+            is RustStructItem -> getStructIcon(element, flags)
+            is RustStructDeclField -> getStructDeclFieldIcon(element, flags)
+            is RustFnItem -> RustIcons.FUNCTION
+            is RustImplMethod -> RustIcons.METHOD
             else -> null
         }
     }
-}
\ No newline at end of file
+
+    private fun getEnumIcon(element: RustEnumItem, flags: Int): Icon? {
+        if ((flags and Iconable.ICON_FLAG_VISIBILITY) == 0)
+            return RustIcons.ENUM;
+
+        val parent = element.parent?.parent as RustItem?
```

Comment:
PSI is a Program Structure Interface: http://www.jetbrains.org/intellij/sdk/docs/reference_guide/custom_language_support/implementing_parser_and_psi.html

PSI is indeed generated by [Grammar Kit](https://github.com/JetBrains/Grammar-Kit) from the Rust [grammar](https://github.com/alexeykudinkin/intellij-rust/blob/master/src/org/rust/lang/core/grammar/rust.bnf), but it is possible to extend PSI elements.


If you want just to add a method to some particular PSI element, and don't need dynamic dispatch, polymorphism and a different type (you are not going to use `is` casts and visitors) you can write an extension method: [example](https://github.com/alexeykudinkin/intellij-rust/blob/master/src/org/rust/lang/core/psi/util/RustPsiUtil.kt#L23)


But if you do need dynamic dispatch, you'll have to go through some tricky inheritance hops. For each PSI element, and inerface and an implementation class are generated. You need to extend the interface, and then to extend imlementation class to implement the extended interface. You do it like this:

* declare an interface, which will be extended by PSI element interface: https://github.com/alexeykudinkin/intellij-rust/blob/master/src/org/rust/lang/core/resolve/scope/RustResolveScope.kt#L9

* ask gramar kit to add this interface to some PSI element: https://github.com/alexeykudinkin/intellij-rust/blob/master/src/org/rust/lang/core/grammar/rust.bnf#L145

* write a "mixin" which implements the interface: https://github.com/alexeykudinkin/intellij-rust/blob/master/src/org/rust/lang/core/psi/impl/mixin/RustBlockImplMixin.kt#L12 The mixin will be the direct  superclass of an actuall class that implements PSI element.

* ask grammar kit to use your mixin: https://github.com/alexeykudinkin/intellij-rust/blob/master/src/org/rust/lang/core/grammar/rust.bnf#L146